### PR TITLE
Clear-ComputeNode.ps1

### PR DESCRIPTION
**Pull request by jacek-izykowski**

### CODE REVIEW

File: Clear-ComputeNode.ps1 [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            Stop-Service $AgentServiceName -ErrorAction Stop
+            sc.exe delete $AgentServiceName
+            if ($LASTEXITCODE -ne 0) {
+                throw "sc.exe failed to delete service."
+            }
+        }
+    } catch {
+        $ErrorMessage += $_.Exception.Message
+        Write-Host "    This step failed with the following error message: $ErrorMessage"
+        Write-Host "    Trying to continue, but final result is unpredictable."
```
Comment (m-kostrzewa) [inline Clear-ComputeNode.ps1]: We should probably wrap this error reporting in a function

File: Clear-ComputeNode.ps1 [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    }
+}
+
+function Remove-AgentService {
```
Comment (m-kostrzewa) [inline Clear-ComputeNode.ps1:68]: All `Remove-*Service` functions are really similar. Maybe we can remove this triplication?

File: README.md [hunk 3]
```diff
@@ -34,3 +34,51 @@ system:
...
+Arguments `AdapterName`, `ForwardingExtensionName` and `VMSwitchName` refer to: name of physical interface controller by vRouter, name of forwarding extension and name of virtual switch created by vRouter.
+
+## Clear-RemoteComputeNodes.ps1
+
+This script runs clean up script on remote compute nodes described by comma separated list of addresses given in `Addresses` argument (e.g. `"10.0.19.5, 10.0.19.83"`)
+
+Example invocation:
+```
+.\Clear-RemoteComputeNodes.ps1 `
+    -ScriptFileName "Clear-ComputeNode.ps1" `
```
Comment (m-kostrzewa) [inline README.md:70]: Hmm, maybe let's rename `Clear-RemoteComputeNodes.ps1` to `Invoke-ScriptInRemoteSessions.ps1`, and use Powershell splats to pass on arbitrary parameters?
https://technet.microsoft.com/en-us/library/gg675931.aspx
```
$parms = @{'class'='Win32_BIOS';
		'computername'='SERVER-R2';
           'filter'='drivetype=3';
           'credential'='Administrator'
          }
Get-WmiObject @parms
```

Comment (m-kostrzewa) [reply README.md:70]: Then, we can provide an example "you can run arbitrary scripts, like Clear-ComputeNode..."

File: README.md [hunk 4]
```diff
@@ -34,3 +34,48 @@ system:
...
+    -AdapterName "Ethernet1" `
+    -ForwardingExtensionName "vRouter forwarding extension" `
+    -VMSwitchName "Layered Ethernet1" `
+    -ConfigAndLogDir "C:\ProgramData\Contrail" `
+    -InstallationDir "C:\Program Files\Juniper Networks"
+```
+
+All arguments are optional. Default values are demonstrated in the example above.
+
+Arguments `AdapterName`, `ForwardingExtensionName` and `VMSwitchName` refer to: name of physical interface controller by vRouter, name of forwarding extension and name of virtual switch created by vRouter.
```
Comment (m-kostrzewa) [inline README.md]: `s/controller/controlled`

> virtual switch created by vRouter

`used by vRouter` or `where vRouter is installed`

File: README.md [hunk 5]
```diff
@@ -34,3 +34,48 @@ system:
...
+
+## Invoke-ScriptInRemoteSessions.ps1
+
+This script runs a script (specified by `ScriptFileName`) on remote compute nodes described by comma separated list of addresses given in `Addresses` argument (e.g. `"10.0.19.5, 10.0.19.83"`).
+
+Example invocation:
+```
+.\Clear-RemoteComputeNodes.ps1 `
```
Comment (m-kostrzewa) [inline README.md]: Invoke-ScriptInRemoteSessions.ps1

File: Invoke-ScriptInRemoteSessions.ps1 [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                    -ScriptFileName $Using:ScriptFileName `
+                    -Parameters $Using:ArgumentsToPass
+            }
+        } catch {
+            Write-Host "ERROR: $($_.Exception.Message)"
+        }
+    }
+}
+
+$Sessions = New-Sessions -Addresses $Addresses.Split(",")
```
Comment (m-kostrzewa) [inline Invoke-ScriptInRemoteSessions.ps1:94]: In example you give "10.0.19.5, 10.0.19.83", but for second IP address, there would be a leading whitespace. Does this break anything?

File: Invoke-ScriptInRemoteSessions.ps1 [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+Invoke-ScriptInSessions -ScriptFileName $ScriptFileName -Sessions $Sessions
+Close-Sessions -Sessions $Sessions
```
Comment (m-kostrzewa) [inline Invoke-ScriptInRemoteSessions.ps1:96]: See my other comment about failing to open session being fatal. Then, I think we should
```
try {
new-sessions
invoke...
} finally {
foreach ($s in $sessions) {
  if ($s is open or whatever) { $s.close() }
}
```

File: Invoke-ScriptInRemoteSessions.ps1 [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    return $Sessions
+}
+
+function Close-Sessions {
+    Param ([Parameter(Mandatory=$true)] [System.Management.Automation.Runspaces.PSSession[]] $Sessions)
+
+    Write-Host "Closing Powershell sessions..."
+    foreach ($Session in $Sessions) {
+        Remove-PSSession -Session $Session
```
Comment (m-kostrzewa) [inline Invoke-ScriptInRemoteSessions.ps1:48]: We need to make sure if session is open before closing it, don't we?

File: Invoke-ScriptInRemoteSessions.ps1 [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        try {
+            if ($IndividualCredentials -eq $true) {
+                $Credential = Get-Credential -Message "Enter credentials for $Address"
+            }
+            $PSO = New-PSSessionOption -MaxConnectionRetryCount $RetryCount -OperationTimeout $Timeout -ErrorAction Stop
+            $Session = New-PSSession -ComputerName $Address -Credential $Credential -SessionOption $PSO -ErrorAction Stop
+            $Sessions += $Session
+            Write-Host "Session created: $($Session.ComputerName)"
+        } catch {
+            Write-Host "ERROR: Failed to create session for remote compute node: $Address. Skipping this compute node."
```
Comment (m-kostrzewa) [inline Invoke-ScriptInRemoteSessions.ps1:36]: Maybe this should be a fatal error?

File: Invoke-ScriptInRemoteSessions.ps1 [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    [Parameter(Mandatory = $false)] [String] $Addresses = "127.0.0.1",
+    [Parameter(Mandatory = $false)] [Switch] $IndividualCredentials,
+    [Parameter(Mandatory = $false, ValueFromRemainingArguments = $true)] $ArgumentsToPass
+)
+
+function New-Sessions {
+    Param (
+        [Parameter(Mandatory=$true)] [String[]] $Addresses,
+        [Parameter(Mandatory=$false)] [Int] $RetryCount = 10,
+        [Parameter(Mandatory=$false)] [Int] $Timeout = 300000
```
Comment (m-kostrzewa) [inline Invoke-ScriptInRemoteSessions.ps1]: this is 5 minutes? way too long, we probably need to aim for like 5 seconds. Please include units (ms / seconds, whatever) in variable name.

File: Clear-ComputeNode.ps1 [hunk 11]
```diff
@@ -0,0 +1,202 @@
+Param (
+    [Parameter(Mandatory = $false)] [String] $AdapterName = "Ethernet1",
+    [Parameter(Mandatory = $false)] [String] $ForwardingExtensionName = "vRouter forwarding extension",
+    [Parameter(Mandatory = $false)] [String] $VMSwitchName = "Layered Ethernet1",
+    [Parameter(Mandatory = $false)] [String] $ConfigAndLogDir = "C:\ProgramData\Contrail",
+    [Parameter(Mandatory = $false)] [String] $InstallationDir = "C:\Program Files\Juniper Networks"
+)
+
+function Invoke-ScriptBlockAndPrintErrors {
```
Comment (m-kostrzewa) [inline Clear-ComputeNode.ps1]: Maybe `Invoke-ScriptBlockAndPrintExceptions` would be better?

File: Clear-ComputeNode.ps1 [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    $Proc = Get-Process $ProcessName -ErrorAction SilentlyContinue
+    if ($Proc) {
+        $Proc | Stop-Process -Force -PassThru | Wait-Process -ErrorAction Stop
+    }
+}
+
+function Remove-HNSNetworks {
+    Write-Host "Cleaning HNS state..."
+    Invoke-ScriptBlockAndPrintErrors {
+        Get-NetNat | Remove-NetNat -Confirm:$false
```
Comment (m-kostrzewa) [inline Clear-ComputeNode.ps1]: Removing NetNat should be a separate function I think. It's only indirectly related to HNS network created by docker upon startup.

File: Invoke-ScriptInRemoteSessions.ps1 [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    )
+
+    if ($IndividualCredentials -eq $false) {
+        $Credential = Get-Credential -Message "Enter common credentials for compute nodes"
```
Comment (m-kostrzewa) [inline Invoke-ScriptInRemoteSessions.ps1:21]: What do you think about passing Credential object to this script and assuming they are the same? If we don't pass Credential object, then ask the user for each one.

**Review by m-kostrzewa — CHANGES_REQUESTED** [review]

File: Invoke-ScriptInRemoteSessions.ps1 (hunk 8 above)
Comment (jacek-izykowski) [reply Invoke-ScriptInRemoteSessions.ps1:48]: We don't. Remove-PSSession can be called even for sessions that have already been removed.

File: Invoke-ScriptInRemoteSessions.ps1 (hunk 9 above)
Comment (jacek-izykowski) [reply Invoke-ScriptInRemoteSessions.ps1:36]: Why? Let's run the script for valid sessions.

**Review by m-kostrzewa — APPROVED** [review]

**Pull request merged**